### DISCUSSION TITLE
Trpc 11

### DISCUSSION THREAD

**Discussion by jwtong** [omar-dulaimi/trpc-shield#25] [Ideas]
Seems like only change needs to be for createRouter, rawInput is now an async function getRawInput().

**Comment by omar-dulaimi** [comment]
Done, and it is exactly the change you identified.

Shipped in **3.0.0**. `@trpc/server` is now `^11.18.0`, and the middleware takes `getRawInput` off the options and calls it:

```ts
// src/generator.ts
const { next, ctx, type, path, input, getRawInput } = opts;
const rawInput = getRawInput();
```

Your rule signature is unchanged, so `rawInput` still arrives as a value rather than a function and nothing in your rules needs touching. You spotted the one thing that actually had to move.

One thing to know before upgrading, because it is more serious than the v11 support and affects everyone on 1.x or 2.x: **the published package was unloadable**. `package.json` declared `"type": "module"` while the build emitted CommonJS, so both `require()` and `import` failed on every version from 1.0.0 (July 2025) through 2.0.1. The test suite never caught it because the tests import from `src/`, not from the built output.

3.0.0 dual-publishes properly, with CommonJS and ESM behind an `exports` map, and there is now a check that packs the tarball, installs it into an empty directory and loads it both ways before publish. I drove a real tRPC v11 router through the installed package in both module systems to confirm.

Thanks for the report, and sorry it sat this long.
